Device Plugin Protobuf API

RenaudWasTaken opened the pull request:
**What this PR does / why we need it:**
This implements the Device Plugin API

- Design document: kubernetes/community#695
- PR tracking: [kubernetes/features#368](https://github.com/kubernetes/features/issues/368#issuecomment-321625420)

Special notes for your reviewer:

First proposal submitted to the community repo, please advise if something's not right with the format or procedure, etc.
@vishh @derekwaynecarr

**Release note:**
```
NONE
```

vishh commented:
@jiayingz what do you think about the API?

Hui-Zhi commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,65 @@
+
+	// HeartbeatOk means that the heartbeat succeeded
+	HeartbeatOk = "Heartbeat ok"
+	// HeartbeatOk means that the heartbeat did not succeed
```
It's better if we only have HeartbeatOk and HeartbeatFailed, if we have HeartbeatFailed returned, we also can get a message about the details.

RenaudWasTaken replied:
That makes a lot of sense!
I actually don't think an error should ever be returned and that in case of failure the device plugin should always register himself

jiayingz commented:
/assign @jiayingz

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+	repeated DeviceSpec spec = 1;
+}
+
+message DeviceSpec {
```
Wonder whether we should call this DeviceRuntimeInfo or some other name. DeviceSpec sounds more like some static device properties.

RenaudWasTaken replied:
Renamed to DeviceRuntimeSpec

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+*/
+message Device {
+	string kind = 1;
+	string name = 2;
```
s/name/id?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+
+	map<string, string> properties = 4; // Could be [1, 1.2, 1G]
+
+	DeviceSpec spec = 5;
```
We also have this field in AllocateResponse. That seems redundant. Suggest we either keep it here or in AllocateResponse, but not both.

RenaudWasTaken replied:
I though Allocate was supposed to be optional.
If it's not activated and this field isn't in the Device what does the Device Plugin do ?

jiayingz replied:
I was thinking that we could make Allocate optional if it is the only API exported from device plugin service. But given that we have decided to export ListDevice from device plugin service, there won't be an easy way to make Allocate optional. I am now fine with keeping DeviceRuntimeInfo in Allocate response and skip that from ListDevice.

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+*}
+*/
+message Device {
+	string kind = 1;
```
The kind field seems to match the resourcename field I mentioned earlier. Should we move it to RegisterRequest?

RenaudWasTaken replied:
If we decide for Device Plugin to support multiple Kind in the future the Kind here will make it easier for the API to evolve

jiayingz replied:
We can add that field when the future use cases come out :).

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+
+message DeviceSpec {
+	// List of environment variable to set in the container.
+	repeated KeyValue envs = 1;
```
Instead of defining your own KeyValue struct, how about using map<string, string> here?

RenaudWasTaken replied:
This is a copy of the CRI struct.
I'd rather keep things the same instead of having the same field with a different Type

jiayingz replied:
I am ok to leave it this way. Just feel a bit unnecessary to have this type definition.

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+/* E.g:
+* struct Device {
+*    Kind: "nvidia-gpu"
+*    Name: "GPU-fef8089b-4820-abfc-e83e-94318197576e"
```
Could you also add state field in the comment example?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+	string name = 2;
+	string state = 3;
+
+	map<string, string> properties = 4; // Could be [1, 1.2, 1G]
```
The comment in this line doesn't seem to provide additional value. Can we omit that?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_unrecognized_all) = false;
+
+
```
Could you add some comment here on who is supposed to export the Registration service (i.e., Kubelet) and what would happen next after Kubelet receives a Registration rpc call?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+
+
+service Registration {
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
```
Could you add some comment here explaining what this rpc does, when Register may fail, and what would happen when registration succeed or fail?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+	rpc Register(RegisterRequest) returns (RegisterResponse) {}
+}
+
+service DevicePlugin {
```
Similar here, could you add some comment explaining who is supposed to export this service and what each rpc API does?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,63 @@
+	HeartbeatOk = "Heartbeat ok"
+	// HeartbeatFailure means that the heartbeat did not succeed
+	HeartbeatFailure = "Heartbeat Failure"
+
```
Do we still need Heatbeat related strings?

RenaudWasTaken replied:
Nope removing. Good catch :)

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,63 @@
+	// Version is the API version
+	Version = "alpha"
+	// DevicePluginPath is the folder the Device Plugin is expecting sockets to be on
+	DevicePluginPath = "/var/run/kubernetes/"
```
Maybe add some comment on that only privileged pod would have access to this directory?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,63 @@
+	DevicePluginPath = "/var/run/kubernetes/"
+	// KubeletSocket is the path the Kubelet registry socket
+	KubeletSocket = DevicePluginPath + "kubelet.sock"
+
```
Can we leave the rest of the const string definitions to later PRs where they are used? That would provide better context on how they will be used.

jiayingz reviewed (commented):
Thanks a lot for the great work, Renaud. The overall design looks good. Here are my initial comments.

RenaudWasTaken commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+}
+
+message Mount {
+	string name = 1;
```
An example here doesn't seem to provide additional value :)
If you don't know what a mount is you're probably not looking at the right API.

I understand though that the name might be contentious but as commented higher I'd rather keep the fields the same as the CRI spec

RenaudWasTaken commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,90 @@
+}
+
+message AllocateRequest {
+	repeated Device devices = 1;
```
Mostly to stay consistent across the API. This also allow us to not have to update this if we allow for multiple Device Kind per Device Plugin.

It's also a lot cleaner because we aren't depending on the type of ID

jiayingz replied:
But this seems require extra work for Kubelet to prepare this request instead of just using simple device id. Also, if Device struct grows big in the future, which is quite likely to happen for more efficient resource scheduling, we don't want to add all those fields here because proto exchange does have its own performance cost. I would prefer to using simple id string here instead of using complicated struct for unclear future use cases.

RenaudWasTaken replied:
> But this seems require extra work for Kubelet to prepare this request instead of just using simple device id. 

Actually it's probably the opposite since Kubelet is going to store Device Object

> Also, if Device struct grows big in the future, which is quite likely to happen for more efficient resource scheduling, we don't want to add all those fields here because proto exchange does have its own performance cost. I would prefer to using simple id string here instead of using complicated struct for unclear future use cases

I don't think that this should be a concern unless we expect properties to be a few Megabyte large.
Can you give an actual example ?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,118 @@
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices sent
+message AllocateRequest {
+	repeated Device devices = 1;
```
Can we just pass device id here instead of the whole Device struct? A device should be uniquely identified by its device id within a device plugin, so it is really unnecessary to pass the whole Device struct. I know you feel using device id would require a few more lines on the Kubelet side, but I feel internal implementation complexity is different from API complexity. If we use the whole Device struct here, we may need to deal with more complexities during upgrade when we change Device struct in the future.

RenaudWasTaken replied:
Here are my arguments:
- It's more consistent to have the API deal only with Devices
- It's actually more of a burden on both side of the Kubelet or Device Plugin because everyone is dealing with Devices except for this specific call
- It makes the AllocateRequest depend explicitly on the ID of the device as well as the ID type
- If we decide that a Device is not only identified by it's ID but also it's type / Kind / Resource Name (e.g: GPU) then this message is irrelevant

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,118 @@
+*}
+*/
+message Device {
+	string resourceName = 1;
```
Can we omit resourceName here given that it is already included in RegisterRequest? If we include this field in both places, Kubelet side needs to detect and handle the case that a device plugin sends a Device struct with non-existing or conflicting resourceName. To detect this, Kubelet side needs to maintain some kind of mapping from device plugin connection to resource_name. So keeping this field in Device struct doesn't seem to provide extra value but just adds more error cases to worry about. If we don't include this field here, the API contract is simpler. Each device is uniquely identified by its id within a device plugin. Kubelet side maintains some kind mapping from device plugin connection to resourceName, but only needs to detect possible conflicting resource_name during registration.


[10:12] 
Similarly, during Allocate, Kubelet only needs to send device ids to device plugin. It doesn't need to send the whole Device struct.

RenaudWasTaken replied:
Here are my arguments:
- A device does not mean anything without it's Kind
- > Each device is uniquely identified by its id within a device plugin
  Exactly the point above, the actual ID of a device is ID + ResourceName

- After multiple implementation of the Device Manager (Kubelet side) I've always had to use the ResourceName whenever I was using the ID.
Having every function call to a Device be followed by the ResourceName should be a strong marker that maybe the Device struct needs to include the Resource Name

- Also I really don't think that we should concern ourselves with one more testcase / error handling case and start getting this API done

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,106 @@
+	// Name of the Device advertised
+	string resourceName = 3;
+	// The labels to advertise at the node level
+	map<string, string> labels = 4;
```
As we discussed in the PR doc, can we remove the labels field for now?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,106 @@
+}
+
+// Allocate is called during container creation so that the Device
+// Plugin can run device specific operations on the Devices sent
```
This sentence seems incomplete?

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,106 @@
+	string ID = 2;
+	string state = 3;
+
+	DeviceRuntimeSpec spec = 5;
```
Suggest we leave out DeviceRuntimeSpec from Device message for now, since we don't have any immediate plan on using it.

RenaudWasTaken replied:
Removing it.
Still think it's incompatible to do that and have the Allocate phase optional :)

jiayingz replied:
Thanks! To make allocate phase optional, we need extra information passed to Kubelet. Since we don't have that yet, probably just don't support optional Allocate for now.

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,32 @@
+	// DevicePluginPath is the folder the Device Plugin is expecting sockets to be on
+	// Only privileged pods have access to this path
+	DevicePluginPath = "/var/run/kubernetes/"
+	// KubeletSocket is the path the Kubelet registry socket
```
s/the path/the path of?

jiayingz reviewed (commented):
Looks good overall. Only a few comments and I think most of them we already agree on.

RenaudWasTaken commented:
Addressed @jiayingz's comments.
Can someone allow testing ? @vishh maybe ?

jiayingz commented:
FYI, this PR lgtm on my side.

cmluciano commented:
/ok-to-test

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+	// ListWatch returns a stream of List of Devices
+	// Whenever a Device state change or a Device disapears, ListWatch
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
```
Would `ListAndWatch` be more appropriate?

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+	// returns the new list
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+
+	// Allocate is called during container creation so that the Device
```
Allocate is expected to be called during pod creation since allocation failures for any container would result in pod startup failure.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+
+	// Allocate is called during container creation so that the Device
+	// Plugin can run device specific operations and instruct Kubelet
+	// of the steps to make the Device available in the container
```
Kubelet does not take steps, rather kubelet exposes additional artifacts in a pod's environment as directed by the plugin.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+message RegisterRequest {
+	// Version of the API the Device Plugin was built against
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
```
Is it name or path? This API is essentially exposing a named socket right?

RenaudWasTaken replied:
It's the name of the socket.

vishh replied:
What's the path to the named socket?

RenaudWasTaken replied:
path.Join(DevicePluginPath, endpoint)

vishh replied:
That's not obvious from the API Spec. How would someone new reading this API know how to implement a plugin?

RenaudWasTaken replied:
We can add a comment to make that more obvious.

vishh replied:
What is the drawback for passing the fully qualified path?

jiayingz replied:
For Kubelet to connect to device plugin through this path, it needs to be full host path. Ideally, we want device plugin to be unaware of this host mount information. Since we already want kubelet to enforce that all device plugin sockets to be under DevicePluginPath for access control, it seems better to omit the DevicePluginPath in API.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+	// Name of the unix socket the device plugin is listening on
+	string endpoint = 2;
+	// Name of the Device advertised
+	string resourceName = 3;
```
This is the `schedulable resource name`, not just a `name`.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+}
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
```
Fix this comment.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+
+message RegisterResponse {
+	// Minimum version the Kubelet API supports.
+	bool success = 1;
```
@jiayingz Why is this field necessary?

jiayingz replied:
This is mostly to give a clear signal on whether registration succeeds or not. Otherwise, we will need to rely on implicit interpretation on the special error message like "SUCCESS". Personally I prefer to having explicit signal, but I am fine to omit this field if the convention in Kubernetes is to embed this in the error message.

vishh replied:
isn't success already conveyed by the fact that that grpc returned success? https://godoc.org/google.golang.org/grpc/codes

jiayingz replied:
It is true that we can use grpc success as the signal. Say in case grpc succeed, do we still need to check error field? I.e., how should we define registration success? Should it be grpc succeeds, or grpc succeeds and error msg empty, or grpc succeeds and error msg being "SUCCESS"? The first option seems fine to me, i.e., grpc succeeds means registration succeeds and error field is only for debugging purpose in case grpc failure. If this sounds a reasonable option, can we remove the success field and add comment that device plugin can use error field for debugging purpose in case registration grpc failure?

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,105 @@
+}
+
+// The list to be added to the CRI spec
+message DeviceRuntimeSpec {
```
Why isn't a device file included? Device ID are opaque strings right? How will kubelet know what device file to expose unless the plugin instructs it? 

On that note, doesn't the DeviceRuntimeSpec have to be per device ID?

RenaudWasTaken replied:
> Why isn't a device file included? How will kubelet know what device file to expose unless the plugin instructs it?

It's the Mount struct.

> On that note, doesn't the DeviceRuntimeSpec have to be per device ID?

Yes, is your point that we should have the device ID in the RuntimeSpec instead of having the same order as the one given by allocate ?

vishh replied:
Mount is different from devices. Devices for example require cgroup updates. Let's split them.

> Yes, is your point that we should have the device ID in the RuntimeSpec instead of having the same order as the one given by allocate ?

This is an option. I was suggesting something as follows:

```
message AllocateResponse {
  repeated DeviceSpec device_specs;
}

message DeviceSpec {
  string id = 1;
  map<string, string> envs = 2;
  repeated Mount mounts = 3;
}
```

RenaudWasTaken replied:
> Mount is different from devices. Devices for example require cgroup updates. Let's split them.

I agree that they require cgroup update. I was thinking of looking if the host path starts with /dev
If we want to split them we can do that

vishh replied:
No hacks in the APIs please.

RenaudWasTaken replied:
> I am also curious whether we should consider device allocation failure as transient or permanent. 

 IMHO it's the job of ListWatch to mark a device unhealthy. If an Allocate fails, it's the job of the Device Plugin to mark the devices unhealthy though ListWatch.

> Some question related to the AllocateResponse API is what would happen if device plugin fails to "allocate" some of the requested devices (...) but succeeds on others

We can have a retry policy with random devices, but in general Allocate should not be a vector of communicating a Device Error / Unhealthy.

> Say if a gpu device fails the simple test we will run before each allocation, how can we recover the device?

Usually when a GPU fails the driver also fails and the machine crashes.
You will probably need human intervention here

jiayingz replied:
Say if Kubelet sends device plugin an allocation request for dev1 and dev2. Allocation on dev1 succeeds but allocation on dev2 fails. Should device plugin a) send a ListWatch update and fail the Allocation request, or b) send a ListWatch update and response an AllocationResponse with dev1 only, or c) response an AllocationResponse with dev1 only without ListWatch update? From your response, I guess you are suggesting a) or b)? On the Kubelet side, upon Allocation failure, should it retry for up to some specified times or should it fail the allocation? I am fine with your proposed approach, but we need to comment about this clearly so that we have a consistent behavior on different device plugin implementations. And more ideally, it would be nice if we can have a device plugin lib that does this.

RenaudWasTaken replied:
IMHO a) seems like a correct architectural solution.

> should it retry for up to some specified times or should it fail the allocation.

I'm not against a retry policy. I'm not sure how it would work with the CRI process which I think has a timeout (I might be wrong on that part).

jiayingz replied:
I am fine with option a) for 1.8. We can even skip retry part on Kubelet. Can we comment with option a) above AllocateResponse to set clear expectations for device plugin implementation? Btw, we probably need to revisit this part in future versions when we need to support other types of devices that may have different device failure modes. But for 1.8, nvidia gpu is likely the only device we will test against.

RenaudWasTaken replied:
I'm fine with this :)
Are the current comments enough ?

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,32 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Change the year in all files.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,32 @@
+	Unhealthy = "Unhealthy"
+
+	// Version is the API version
+	Version = "alpha"
```
Why is the version `alpha`? Why is it not something more structured like `0.1` for example? @jiayingz have you thought through versioning already?

jiayingz replied:
I think we can start by requesting exact version match between kubelet and device plugin, and only increases API version with incompatible proto change. Agree number versioning is more appropriate.

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,32 @@
+	Version = "alpha"
+	// DevicePluginPath is the folder the Device Plugin is expecting sockets to be on
+	// Only privileged pods have access to this path
+	DevicePluginPath = "/var/run/kubernetes/"
```
Why is this the path? Is this a standard path already across the node?

RenaudWasTaken replied:
> Why is this the path?

This is a comment that came up multiple time in the design document.
This was my answer:

> That's mostly a placeholder until we find a "standard path" do you think there is a more appropriate path ?

vishh replied:
Can you mention this in the PR?

RenaudWasTaken commented:
/test pull-kubernetes-unit

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,114 @@
+	// Mounts for the container.
+	repeated Mount mounts = 3;
+
+	repeated Mount devices = 4;
```
What is the last field intended to be used?

RenaudWasTaken replied:
To "mount" devices in the container

vishh replied:
There is no "mount" happening. What's happening is an `mknod`. Can we have an explicit message for Devices?

RenaudWasTaken replied:
Done

RenaudWasTaken commented:
/test pull-kubernetes-bazel

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,127 @@
+}
+
+// Device specifies a host device to mount into a container.
+message Device {
```
This would conflict with the Device message defined later, right? Maybe call it RuntimeDevice instead?

RenaudWasTaken replied:
Yep sorry....
I didn't even regen the API :(

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,127 @@
+	repeated DeviceSpec devices = 4;
+}
+
+// Device specifies a host device to mount into a container.
```
s/Device/DeviceSpec?

RenaudWasTaken commented:
/test pull-kubernetes-verify

RenaudWasTaken commented:
/test pull-kubernetes-verify

RenaudWasTaken commented:
/test pull-kubernetes-verify

jiayingz commented:
On Tue, Aug 8, 2017 at 12:16 AM, Renaud Gaubert <notifications@github.com>
wrote:

> *@RenaudWasTaken* commented on this pull request.
> ------------------------------
>
> In pkg/kubelet/apis/device-plugin/v1alpha1/api.proto
> <https://github.com/kubernetes/kubernetes/pull/49342#discussion_r131836750>
> :
>
> > +	repeated string devicesIDs = 1;
> +}
> +
> +message AllocateResponse {
> +	repeated DeviceRuntimeSpec spec = 1;
> +}
> +
> +// ListWatch returns a stream of List of Devices
> +// Whenever a Device state change or a Device disapears, ListWatch
> +// returns the new list
> +message ListWatchResponse {
> +	repeated Device devices = 1;
> +}
> +
> +// The list to be added to the CRI spec
> +message DeviceRuntimeSpec {
>
> I'm fine with this :)
> Are the current comments enough ?
>
To make it clear, can we comment the option we take (i.e., option a)
explicitly on top of AllocateResponse?

Jiaying

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/49342#discussion_r131836750>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AcIZlAgW2vGYC22qP-wl142r3Chn4A48ks5sWAtEgaJpZM4Oe2Vj>
> .
>

RenaudWasTaken commented:
@jiayingz done!

jiayingz commented on pkg/kubelet/apis/device-plugin/v1alpha1/constants.go:
```diff
@@ -0,0 +1,33 @@
+	// DevicePluginPath is the folder the Device Plugin is expecting sockets to be on
+	// Only privileged pods have access to this path
+	// Note: Placeholder until we find a "standard path"
+	DevicePluginPath = "/var/run/kubernetes/"
```
Should we use "/home/kubernetes/device-plugins/" here instead? I think /var/run/kubernetes is what used at apiserver. Kubelet currently doesn't use this directory but seems to keep most state either under /home/kubernetes or /var/lib/kubelet. If we want kubelet to create a directory under /var/run, we probably want to use /var/run/kubelet/device-plugins?

RenaudWasTaken replied:
I'm fine with either :)
@vishh any opinion ?

jiayingz replied:
FYI, just checked with cri folks. A path under /var/lib/ seems better because /home/kubernetes is mostly for staging. I guess device plugin unix socks can live under /var/lib/device-plugins/ and kubelet side device-plugin registration sock and checkpoint data can live under /var/lib/kubelet/device-plugin/?

RenaudWasTaken replied:
I would rather have the unix socket live in the same folder.
What reasons do we have to put them in two different paths ?

It makes the pod spec as well as the general spec more complex.

jiayingz replied:
Configuration complexity is a fair point. I am fine to start with the single /var/lib/kubelet/device-plugins directory.

RenaudWasTaken commented:
/test pull-kubernetes-verify

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-etcd3
/test pull-kubernetes-verify

vishh commented:
Why were the generation scripts pulled out - https://github.com/kubernetes/kubernetes/pull/50401 ?
I'd like to see all the scripts in the same PR. @RenaudWasTaken can you also add your scripts to `hack/update-all.sh` and `hack/make-rules/verify.sh` inorder for them to be run as part of presubmits?

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,124 @@
+	// PATH = path.Join(DevicePluginPath, endpoint)
+	string endpoint = 2;
+	// Schedulable resource name
+	string resource_name = 3;
```
It is undecided yet what the format for this name should be. As of now it's expected to be a [DNS Label](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L48). If that changes, we can update this API Spec. Can you mention the naming constraints explicitly?

RenaudWasTaken replied:
Will do

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,124 @@
+// - Allocate allows Device Plugin to run device specific operations on
+//   the Devices requested
+message AllocateRequest {
+	repeated string devicesIDs = 1;
```
What's the format for Device IDs? Can we stick to DNS labels (63 characters) ?

RenaudWasTaken replied:
> What's the format for Device IDs? Can we stick to DNS labels (63 characters) ?

No specific format. I'd rather we don't stick to DNS labels.
This is meant to be a Device Plugin controlled field and not Kubernetes controlled.

In the case of GPUs we will be using UUIDs and I expect other vendors to use other naming schemes

vishh replied:
I want to have some bounds on the memory used by these device IDs in kubelet. 
I don't see a reason for needing more than 63 characters. 
Keep in mind that kubelet also runs on Raspberry PIs :)

RenaudWasTaken replied:
I don't oppose to bounding the size of the field. :)
However I don't want to impose a specific naming scheme on something that is the device plugin's responsibility.
Adding a comment

RenaudWasTaken commented:
> Why were the generation scripts pulled out - #50401 ?

This was a suggestion by @jiayingz to accelerate approval.

jiayingz commented:
@vishh I suggest Renaud to take out the generation scripts, mostly to unblock the subsequent PR reviews that depend on the proto API. I am totally fine to include them back if the test failure problem can be fixed soon. But if that part will take some part, we may consider to pull them into a separate PR.

RenaudWasTaken commented:
> can you also add your scripts to hack/update-all.sh and hack/make-rules/verify.sh inorder for them to be run as part of presubmits?

I noticed that CRI did not add the script (`update-generated-runtime.sh`) to update-all so not sure it's the right time to do it :)

vishh commented:
@RenaudWasTaken To avoid cognitive burden, I'd add the update scripts as well, but with a check where unless the plugin proto files are dirty, the script will be a no-op.

RenaudWasTaken commented:
/test pull-kubernetes-unit

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-etcd3

vishh commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,125 @@
+message Device {
+	// Max length of this field is 63 characters
+	string ID = 2;
+	string health = 3;
```
nit: Add a comment describing the purpose of these fields.

RenaudWasTaken replied:
@vishh suggested otherwise as he found that enums are hard to maintain

vishh commented:
Just one more nit - https://github.com/kubernetes/kubernetes/pull/49342#discussion_r132771674

/lgtm
/approve

mindprince commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,125 @@
+	// Path of the mount on the host.
+	string host_path = 1;
+	// Path of the mount within the container.
+	string mount_path = 2;
```
Why call it `mount_path` here but the similar field is called `container_path` in `DeviceSpec`?

Also, can we have the same ordering of the two fields in both the messages?

RenaudWasTaken replied:
I followed the same pattern as for CRI: -
-
 [Mount](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L152)
- [Device](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto#L544)

But we don't have to follow the same pattern

mindprince replied:
CRI is following the correct pattern. This PR is not.

In CRI, both Mount and Device have:
    string container_path = 1;
    string host_path = 2;

Here, we have different ordering and names for the these fields.

RenaudWasTaken replied:
Right! Thanks for catching this!

vishh commented:
@spxtr can you review the `hack/*` files in this PR?

mindprince commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,125 @@
+// Registration may fail when device plugin version is not supported by
+// Kubelet or the registered resourceName is already taken by another
+// active device plugin. Device plugin is expected to terminate upon registration failure
+service Registration {
```
Nit: It would make this file much more easier to read if you order the definitions like:

service Registration
message RegisterRequest
message Empty

service DevicePlugin
message ListAndWatchResponse
message Device
message AllocateRequest
message AllocateResponse
message DeviceRuntimeSpec
message Mount
message DeviceSpec

RenaudWasTaken commented on pkg/kubelet/apis/device-plugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,125 @@
+	string version = 1;
+	// Name of the unix socket the device plugin is listening on
+	// PATH = path.Join(DevicePluginPath, endpoint)
+	string endpoint = 2;
```
Sure do you have a suggestion ?
@jiayingz suggested it as it abstracts away the concept of sockets

jiayingz replied:
I think we chose unix socket for now. It is possible that we may use some other communication channel in the future. That is why I suggested to use endpoint.

jiayingz commented:
/assign @ixdy

mindprince commented:
Since we are going to call the package `deviceplugin`, let's rename the directory to `deviceplugin` instead of `device-plugin` in both this PR and the implementation PR.

ixdy commented:
how is this different than the preexisting `hack/update-generated-protobuf` scripts? why can't this functionality be folded into that?

RenaudWasTaken commented:
@ixdy looking at `hack/update-generated-protobuf`it is indeed the same.

I followed the same pattern as `hack/update-generated-runtime` and ended up creating a file in `hack/lib/protoc.sh`

Let me update the scripts

RenaudWasTaken commented:
Though looking at it more in details there seem to be a small difference: `update-generated-protobuf-dockerized` does not gofmt the file and does not add the license

@ixdy how is the `update-protobuf` expected to be used if I want to add the license + gofmt?

Am I expected to have:
- `update-device-plugin.sh` which calls into `update-device-plugin-dockerize.sh`
- `update-device-plugin-dockerize.sh` which then calls into `update-protobuf-dockerize.sh` and then formats the file + add the license ?

Also what do you think of the hack/lib/protoc.sh model ?

Commit:
Added Device plugin API

Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>

Commit:
Added script to generate the Device Plugin API

Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>

Commit:
Added script to verify the generated Device Plugin API

Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>

Commit:
Generated files

Signed-off-by: Renaud Gaubert <rgaubert@nvidia.com>

RenaudWasTaken commented:
@mindprince addressed most of your comments

@ixdy after looking into `go-to-proto` more I get the impression that it does not address our use case. 
It seems to me that `go-to-proto` (or `hack/update-generated-proto.sh`) is used to generate protobuf from go files and then generate the go files from the protobuf (basically the model for the K8s API).

However the Device Plugin API uses a different model (similar to the CRI API) where we write the protobuf file and then generate the golang from that protobuf file.

It seems to me that the script model used in `hack/update-generated-runtime.sh` fits our use case which is why I used it.
However, I could not make `hack/update-generated-protobuf.sh` behave in a similar way. Is there a way to adapt it to our use case ?

If not what do you think of the current model ? I plan to update the `hack/update-generated-runtime.sh` in a later PR to use it.

Thanks!

mindprince commented on hack/.golint_failures:
```diff
@@ -235,6 +235,7 @@ pkg/kubelet
 pkg/kubelet/apis
 pkg/kubelet/apis/cri/testing
 pkg/kubelet/apis/cri/v1alpha1/runtime
+pkg/kubelet/apis/device-plugin/v1alpha1
```
deviceplugin instead of device-plugin.

mindprince commented on pkg/kubelet/apis/deviceplugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,127 @@
+	// A unique ID assigned by the device plugin used
+	// to identify devices during the communication
+	// Max length of this field is 63 characters
+	string ID = 2;
```
This should be 1 and health should be 2.

mindprince commented on pkg/kubelet/apis/deviceplugin/v1alpha1/api.proto:
```diff
@@ -0,0 +1,127 @@
+// The list to be added to the CRI spec
+message DeviceRuntimeSpec {
+	string ID = 1;
+
```
unnecessary new line.

rohitagarwal003 reviewed (commented):
Minor comments. API definitions LGTM otherwise.

spxtr commented on hack/lib/protoc.sh:
```diff
@@ -0,0 +1,91 @@
+# exit 1 if it's not the case
+function kube::protoc::check_protoc() {
+  if [[ -z "$(which protoc)" || "$(protoc --version)" != "libprotoc 3."* ]]; then
+    echo "Generating protobuf requires protoc 3.0.0-beta1 or newer. Please download and"
```
Send all of this to stderr. You can wrap it in `( ... ) 1>&2` I think.

spxtr commented:
IIUC this adds a new generated file and a verify script. We're attempting to remove both of these things from the repo. Would it be possible to not check in the generated file but instead create it when we run `make` and also add a [build rule](https://github.com/bazelbuild/rules_go#go_proto_library) for it? This one in particular looks pretty harmless in terms of merge conflicts, but it does add 90 seconds to the presubmit time and now is the best time to do it right.

cc @ixdy, since I'm not sure what the precedence is here.

jiayingz commented:
@spxtr Agree it would work much better to not check in the generated pb.go code but instead generate it during build time. However, I feel such a change is perhaps out of scope of this PR. I am not aware of any such working model inside kubernetes code tree so I am afraid it would require some substantial changes. I think this PR is following the existing proto generation model used by other components like CRI, and actually improves it a bit with some generic script, which seems good enough for the added API proto.

spxtr commented:
Yeah, I looked into that yesterday and I agree that it's outside the scope of this PR.

/approve no-issue

vishh commented:
/lgtm
(again)

jiayingz commented:
/retest

vishh commented:
/lgtm

derekwaynecarr commented:
Excited to see this merge...

RenaudWasTaken commented:
/test pull-kubernetes-e2e-gce-etcd3
